src/main/java/org/opentripplanner/graph_builder/module/osm/naming/SidewalkNamer.java

```diff
@@ -0,0 +1,275 @@
+      // the sidewalk. This is a safety mechanism for sidewalks that snake around the corner,
+      // like https://www.openstreetmap.org/way/1059101564.
+      .filter(group -> group.percentInBuffer > MIN_PERCENT_IN_BUFFER)
+      .max(Comparator.comparingDouble(NamedEdgeGroup::percentInBuffer))
+      .ifPresent(group -> {
+        namesApplied.incrementAndGet();
+        sidewalk.setName(Objects.requireNonNull(group.name));
+      });
+  }
+
+  /**
+   * Compute the length of the group that is inside the buffer and return it as a percentage
+   * of the length of the sidewalk.
+   */
+  private static NamedEdgeGroup computePercentInsideBuffer(
+    CandidateGroup g,
+    Geometry buffer,
+    double sidewalkLength
+  ) {
+    var lengthInsideBuffer = g.intersectionLength(buffer);
+    double percentInBuffer = lengthInsideBuffer / sidewalkLength;
+    return new NamedEdgeGroup(percentInBuffer, g.name);
+  }
+
+  /**
+   * If a single street is split into several edges, each individual part of the street would potentially
+   * have a low similarity with the (longer) sidewalk. For that reason we combine them into a group
+   * and have a better basis for comparison.
+   */
+  private static Stream<CandidateGroup> groupEdgesByName(List<EdgeOnLevel> candidates) {
+    return candidates
+      .stream()
+      .collect(Collectors.groupingBy(e -> e.edge.getName()))
+      .entrySet()
+      .stream()
+      .map(entry -> {
+        var levels = entry
+          .getValue()
+          .stream()
+          .flatMap(e -> e.levels.stream())
+          .collect(Collectors.toSet());
+        return new CandidateGroup(
+          entry.getKey(),
+          entry.getValue().stream().map(e -> e.edge).toList(),
+          levels
+        );
+      });
+  }
+
+  /**
+   * Add a buffer around a geometry that makes sure that the buffer is the same distance (in meters)
+   * anywhere on earth.
+   * <p>
+   * Background: If you call the regular buffer() method on a JTS geometry that uses WGS84 as the
+   * coordinate reference system, the buffer will be accurate at the equator but will become more
+   * and more elongated the farther north/south you go.
+   * <p>
+   * Taken from https://stackoverflow.com/questions/36455020
+   */
+  private Geometry preciseBuffer(Geometry geometry, double distanceInMeters) {
+    try {
+      var coordinate = geometry.getCentroid().getCoordinate();
+      String code = "AUTO:42001,%s,%s".formatted(coordinate.x, coordinate.y);
+      CoordinateReferenceSystem auto = CRS.decode(code);
+
+      MathTransform toTransform = CRS.findMathTransform(DefaultGeographicCRS.WGS84, auto);
+      MathTransform fromTransform = CRS.findMathTransform(auto, DefaultGeographicCRS.WGS84);
+
+      Geometry pGeom = JTS.transform(geometry, toTransform);
+
+      Geometry pBufferedGeom = pGeom.buffer(distanceInMeters, 4, BufferParameters.CAP_FLAT);
+      return JTS.transform(pBufferedGeom, fromTransform);
```

Comment:
Well, we do need the buffer to check if the other WGS84 edge geometries are inside of it but we don't actually do distance calculations with the buffer. Do you think that we should convert all edge geometries to another projection?

Reply:
Yes, it just so happens that we may be able to use the de-projected buffer without introducing errors because the next step happens to be working with relative lengths. But that is a non-obvious special case that's a little tricky to verify, and everything else happening in these methods is basically reliant on working with projected coordinates. The question is: what is the reason for constructing a reverse projection and applying it to every buffer? Does this add complexity and make correctness harder to establish?

Looking at `assignNameToSidewalk` L143-144, after preciseBuffer projects the sidewalk geometry and de-projects the buffered sidewalk geometry, it then calls SphericalDistanceLibrary.length which re-projects that same sidewalk geometry into another, less accurate projection system to find its length. That same sidewalk geometry is then re-constructed and essentially re-projected repeatedly to find distances to other nearby edges.

If you do choose to project every edge you are working with (for finding nearest points or distances between nearest points in a later step) then you don't need to de-project this buffer. In any case you could retain and reuse the projected sidewalk geometry in several places. I'm just not sure what choices you've made on what to project and when, because the current version of the code I'm looking at is still using DistanceOp on WGS84 coordinates followed by SphericalDistanceLibrary.fastDistance.

There is not one right answer to this. It's a series of choices that would ideally be kept consistent with each other. I am just pointing out that it's unusual to perform so many conversions back and forth to slightly different projections. Other than the DistanceOp applied to WGS84, the existing code probably produces the output you're intending. So what I'm pointing out are mainly opportunities for optimization and simplification. 

I'm happy to discuss this week to provide whatever feedback you need. I tried to summarize the "traditional" approach to this in OTP/R5 which is applying a fast, approximate sinusoidal projection "just in time" only where needed. Something slightly different may be appropriate here given that you're retaining and reusing a lot of geometries in a single-purpose index.